contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/gst_encoder_psnr.py

```diff
@@ -0,0 +1,305 @@
+        ).format(
+            GST_LAUNCH_BIN,
+            self._golden_sample,
+            self._color_space,
+            self._codec,
+        )
+
+        if self._mux.upper() in (
+            GStreamerMuxerType.MP4MUX.name,
+            GStreamerMuxerType.MATROSKAMUX.name,
+        ):
+            encode_parser = self._codec_parser_map.get(self._codec)
+            final_pipeline = "{} ! {} ! {} ! filesink location={}".format(
+                base_pipeline,
+                encode_parser,
+                self._mux,
+                self.artifact_file,
+            )
+        elif self._mux.upper() == GStreamerMuxerType.AVIMUX.name:
+            final_pipeline = "{} ! {} ! filesink location={}".format(
+                base_pipeline, self._mux, self.artifact_file
+            )
+        else:
+            raise SystemExit(
+                "Error: Pipeline for '{}' mux not implemented.".format(
+                    self._mux
+                )
+            )
+
+        return final_pipeline
+
+    def _v4l2jpegenc_pipeline_builder(self) -> str:
+        """
+        Build gstreamer pipeline for JPEG encoder
+        """
+        if self._platform == "genio-350":
+            raise SystemExit(
+                "Genio 350 platform doesn't support v4l2jpegenc codec"
+            )
+        # Capture the first frame and save it as jpg file
+        final_pipeline = (
+            "{} filesrc location={} ! decodebin ! videorate !"
+            " video/x-raw,framerate=1/1 ! videoconvert ! "
+            "video/x-raw,format={} ! {} ! filesink location={}"
+        ).format(
+            GST_LAUNCH_BIN,
+            self._golden_sample,
+            self._color_space,
+            self._codec,
+            self.artifact_file,
+        )
+        return final_pipeline
+
+    def build_pipeline(self) -> str:
+        """
+        Build the GStreamer commands based on the platform and codec.
+
+        Returns:
+            str: A GStreamer command based on the platform and
+            codec.
+        """
+        if self._codec in (
+            GStreamerEncodePlugins.V4L2H264ENC.value,
+            GStreamerEncodePlugins.V4L2H265ENC.value,
+        ):
+            return self._264_265_pipeline_builder()
+        elif self._codec == GStreamerEncodePlugins.V4L2JPEGENC.value:
+            return self._v4l2jpegenc_pipeline_builder()
+        else:
+            raise SystemExit(
+                "Error: unknow encoder '{}' be used".format(self._codec)
+            )
+
+
+def main() -> None:
+    args = register_arguments()
+    p = project_factory(args)
+    logging.info("Step 1: Generating artifact...")
+    cmd = p.build_pipeline()
+    # execute command
+    execute_command(cmd=cmd)
+    # TODO: implement the metadata checker
+    # logging.info("\nStep 2: Checking metadata...")
+    logging.info("\nStep 3: Comparing PSNR...")
+    compare_psnr(
+        golden_reference_file=p.psnr_reference_file,
+        artifact_file=p.artifact_file,
```

Comment:
We probably need the `.psnr_reference_file` and `.artifact_file` properties in `PipelineInterface` so that `project_factory` can just return an object of type `PipelineInterface`. This way all psnr related functions can just program against this single interface and not have to worry about what the concrete class is.

Reply:
So something like this :
```py
class PipelineInterface(ABC):
    @abstractmethod
    def build_pipeline(self) -> str:
        pass

    @property
    @abstractmethod
    def artifact_file(self) -> str:
        pass

    @property
    @abstractmethod
    def psnr_reference_file(self) -> str:
        pass
```

Reply:
Example:
```py
import abc


class PipelineInterface(abc.ABC):
    @abc.abstractmethod
    def build_pipeline(self) -> str:
        pass

    @property
    @abc.abstractmethod
    def artifact_file(self) -> str:
        pass

    @property
    @abc.abstractmethod
    def psnr_reference_file(self) -> str:
        pass


class Genio350(PipelineInterface):
    def build_pipeline(self) -> str:
        return "pipeline genio350"

    @property
    def artifact_file(self) -> str:
        return 'artifact'

    @property
    def psnr_reference_file(self) -> str:
        return 'reference'


def example(p: PipelineInterface):
    print(p.build_pipeline())
    print(p.artifact_file)
    print(p.psnr_reference_file)



if __name__ == "__main__":
    example(Genio350())
```

Reply:
add more properties if needed :D

Reply:
Good suggestion!